Improve stdlib support for `int`

dbuenzli opened the pull request:
This PR improves stdlib support for `int` values. It adds an `Int` module to the standard library and deprecates the `int_of_string_opt` function.

Deprecating `int_of_string` and `string_of_int` is a bit more involved on the code base and will be done in separate PRs.

A few comments:

1. Following the [`Bool`](https://github.com/ocaml/ocaml/pull/2010) module proposal this PR includes only the non-raising `Int.of_string` function which corresponds to `int_of_string_opt`. Here again I have nothing against renaming this function to `of_string_opt` if this is desired.
2. Unless I made a mistake the rest of the module should be uncontroversial, it follows the common parts of the `Nativeint`, `Int32` and `Int64` modules. In particular it does not define any operators.
3. I think that all non operators functions acting on `int`s of the `Stdlib` module could/should be deprecated aswell. This is not done by this PR but can be added. These functions are `succ`, `pred` and `abs`.
3. Leaving aside `exn` and `unit` I think this is the last builtin datatype that was missing its own module. I'm sure you'll be glad to eventually be able to write `{Set,Map}.Make (Int)` your `stdlib`-based programs !

dbuenzli committed:
Stdlib: add Int module.

pmetzger commented:
I recognize I may be in the minority, but I really prefer when the default functions return `opt` and those that don't are labeled `_exn` or the like, so I favor the suggested name for `of_string`.

dbuenzli commented:
Appveyor timedout.

yallop commented on stdlib/int.mli:
```diff
@@ -0,0 +1,139 @@
+    [2{^[Sys.int_size - 1]} - 1]. *)
+
+val min_int : int
+(** [min_int] is the smallest reprsentable integer,
```
typo: "reprsentable"

dbuenzli replied:
Thanks !

dbuenzli commented:
I would like to move on this PR. 

It seems no one has a strong objection about the module's functionality and existence. A contentious point though might be 1. here again the choice is between: 

1. Keep it as it is now and accept that non regularity in the `stdlib` (an irregularity becoming more regular as it might share it with the `Bool` module)
2. Rename the `Int.of_string` of this PR into `Int.of_string_opt` (sic). 
3. Add a function `Int.of_string` that raises `Failure`.

While I favour 1. full `Stdlib` regularity would entail 2+3.

This PR deprecates the `int_of_string_opt` function in [this message](https://github.com/ocaml/ocaml/pull/2012#issuecomment-417948544) @gasche expressed concerns about the current rash of `Pervasives` deprecations. If the deprecations of this PR are blocking merge I'm willing to remove them (I personally think it's a good time to do this, if I get a clear sign that it is desirable I'm willing to put up a PR that deprecates `succ`, `pred` and `abs`).

nojb commented:
> Keep it as it is now and accept that non regularity in the stdlib (an irregularity becoming more regular as it might share it with the Bool module)

One reason I can think of for **not** doing this is that it will keep the modules `Int`, `Int32`, `Int64`, `Nativeint` from sharing a common signature with this function. Not sure about the relative importance of this point, though.

dbuenzli commented:
I have rebased this PR to resolve the conflicts. At the same time:

1. Given [this decision](https://github.com/ocaml/ocaml/pull/2010#issuecomment-420606047) by @alainfrisch I have removed the deprecation warning for `int_of_string_opt`.
2. Still kept the commit that replaces uses of `int_of_string_opt` by `Int.of_string` (which only appear in the test suite). So that less work is needed once/if 1. becomes a reality.
3. Stayed behind the `Int.of_string` returns an option design despite @nojb's remark.

alainfrisch commented on stdlib/int.mli:
```diff
@@ -0,0 +1,139 @@
+(** Integer values.
+
+    Integers are {!Sys.int_size} bits wide. All operations are taken
+    modulo 2{^[Sys.int_size]}. They do not fail on overflow.
```
Might be worth mentioning the type uses 2's complement encoding to represent signed integers between min_int and max_int.  This is important to know for bitwise operations and the division.

alainfrisch commented:
If being able to share a common interface with Int32/Int64/... (e.g.  to support algorithms abstracted over those modules) is not a goal, the rationale for exposing e.g. `Int.one` is mostly an aesthetic one, no? (i.e. nobody will use it in practice).  Not saying I'm against it, but the case for it does not very strong.

If being able the share a common (sub-)interface is a goal, then it's indeed unfortunate that the `X.of_string` has a different signature between Int and Int32 (i.e. it's more problematic than having an inconsistency between Bool and Int).  That said, it's not a huge problem, particularly because many algorithms abstracted over a concrete representation of integers won't be responsible for parsing numbers from text themselves (meaning that the common interface won't need to include of_string).

One possibility, which I don't necessarily fully believe in myself, but might be worth discussing would be to add `X.parse: string->X.t option`/`X.print: X.t -> string`.  At the cost of breaking a tradition, and making even recent addition already morally deprecated, this would have the advantage of providing an eventually consistent story with a good migration path.

[EDIT: forgot the `option` above]

dbuenzli commented:
> That said, it's not a huge problem, particularly because many algorithms abstracted over a concrete representation of integers won't be responsible for parsing numbers from text themselves (meaning that the common interface won't need to include of_string).

That's my impression aswell and the reason why I would still like to otherwise maximize similarity with the other integer modules and hence include `Int.one`.

dbuenzli commented:
> One possibility, which I don't necessarily fully believe in myself, but might be worth discussing would be to add `X.parse: string->X.t`/`X.print: X.t -> string`

I'm unconvinced by `X.print` especially since that one is usually not problematic: AFAIR all `to_string` functions of the stdlib are total. 

I'm intrigued by the potential of an eventual total regularization of `of_string` functions via an `M.parse : string -> t option` convention but I think this discussion can be left out of this PR.

alainfrisch commented:
Indeed, X.print was just be to have some symmetry with X.parse.

Ok, let's leave the discussion around X.parse for later.

mshinwell commented:
I think we should strive for the largest reasonable common subset of operations between all of the integer modules (like Base does with `Int_intf`).

As an example, the next version of Flambda makes extensive use of parameterisation over modules representing various kinds of numbers.  As it happens, `Int` is not one of the modules involved, since it represents host-width integers rather than target-width integers.  However I would not be surprised if others have similar applications where `Int` might be included in the parameterisation.

Drup commented:
I agree that I would highly prefer all the `of_string` functions to be regular among IntX/Bool/Float/Others. Both because of functors, but also simply because I prefer having more regular APIs. Having both `of_string` of type `string -> t` and an option version for all these modules seems like the best choice.

dbuenzli commented:
> Having both `of_string` of type `string -> t` and an option version for all these modules seems like the best choice

I think it's better if we entice people to not use raising variants and the best way of doing is to not provide them. I could do 1. (see above) to make the thing more regular but I personally don't find that to be a great name (it should have been `Int.option_of_string` in my opinion) and as far as this is concerned I think going in the `X.parse` direction suggested by @alainfrisch above would provide a good eventual solution.

Drup commented:
I didn't advise a name on purpose (I also think `X.parse` is pretty good), my only point was to ensure the regularity. I understand your point about encouraging people not to use the raising versions, but if we deprecate them, let's do that on all the parsing functions at the same time, instead of having some irregular transitional APIs.

dbuenzli commented:
> but if we deprecate them, let's do that on all the parsing functions at the same time, instead of having some irregular transitional APIs.

The raising ones are already irregular, see for example `bool_of_string`. So what should I do now introduce `Bool.of_string` that raises `Failure` which never existed and this even though we agree we should move away from raising variants ? 

I'm personally rather against this both because we have an opportunity to give the good signal here and because I can see people switching from `bool_of_string` to `Bool.of_string` failing to realize they now need to catch another exception.

Drup commented:
I'm not sure I understand. `bool_of_string` has been around for a while, and has always been raising. If you are talking about the fact that `Bool.of_string` is non-raising .... well, I have the same oppositions there, for exactly the same reasons.

One of the reason some alternative standard libraries are used is because of their very high degree of internal consistency. You are making it worse for the stdlib here. If you want to go forward with removing all the raising variants, and I'm sure many people agree, I propose you make a patch for that, instead of doing that change in various semi-related PRs.

Personally, my opinion would be to add `parse` everywhere, put it front and center in the documentation, and keep `to_string` raising, for efficiency purpose. But I would rather debate that somewhere else where we will do it for the whole stdlib.

dbuenzli commented:
> I'm not sure I understand. `bool_of_string` has been around for a while, and has always been raising.

It raises `Invalid_argument`, not `Failure` on error.

> You are making it worse for the stdlib here. 

Frankly, is it that bad ? 

> If you want to go forward with removing all the raising variants, and I'm sure many people agree, I propose you make a patch for that, instead of doing that change in various semi-related PRs.

I think this misses the point somehow. This PR introduces new functionality, I'm just avoiding in this new functionality a mode of operation that we consider we should move away from. I acknowledge it introduces irregularity, I personally don't see it as a big deal. If others also think strongly about this then I think the current `of_string` of this PR should be changed to `of_string_opt` and that's all, I do not want to introduce new names that use the mode of operation we would like to move away from.

Regarding removing the raising variants I was actually on that track (but I don't think it's necessarily something to be done in a single PR), but somehow the discussion about deprecation and its outcome stopped me on the way.

gasche commented:
I wouldn't support removing existing stdlib functions or changing their types, but personally I'm ok with @dbuenzli's choice of having a more modern type for the more modern versions of `of_int`. I would also be fine with a module only exporting `of_int_opt`, and I agree that introducing more exception-raising `of_int` should not be done solely for the sake of consistency.

Drup commented:
> I acknowledge it introduces irregularity, I personally don't see it as a big deal. If others also think strongly about this then I think the current `of_string` of this PR should be changed to `of_string_opt` and that's all, I do not want to introduce new names that use the mode of operation we would like to move away from.

That's fair. I don't know about others, but I do think having regular APIs is a Very Big Deal. So yes, please do the name change, either to `parse` or to `_opt`.

That being said, I *also* disagree that raising variants should never be used. `int_of_string` is used very commonly in contexts where we know that the element to parse will never fail, such as lexers. In such cases, the little bit of efficiency gained by not having to constantly allocate option wrappers can be important. For instance, Tyre's converter used to be of type `'a -> 'b option`. I eventually changed them to `'a -> 'b` with exceptions and the perf gain was occasionally significant. And no, flambda doesn't bridge that gap (yet ?).

gasche commented:
(Also, to play the devil's advocate, `match .. with exception` makes it much easier to write correct code handling exceptions. We only need exception typing...)

alainfrisch commented:
The current implementation of int_of_string_opt is:
````ocaml
let int_of_string_opt s =
  (* TODO: provide this directly as a non-raising primitive. *)
  try Some (int_of_string s)
  with Failure _ -> None
`````

Perhaps it's actually best to leave it like that.  A clever inliner could then optimize:
````ocaml
match int_of_string_opt s with
| Some x -> ...
| None -> ...
````
into:
````ocaml
match int_of_string s with
| x -> ...
| Failure _-> ...
````
(Can flambda do that?)

dbuenzli commented:
> The current implementation of int_of_string_opt is:
> 
> ```ocaml
> let int_of_string_opt s =
>   (* TODO: provide this directly as a non-raising primitive. *)

I actually had the plan to work on this after these things are in. So if you really think it's not worth it please say.

dbuenzli commented:
So it may be time to sum up a little bit the larger picture on these `of_string` interface regarding the new `Int` and `Bool` module which may have been lost in the discussions. Here are a few possibilites, feel free to add your own. Basically it seemed to me that we had agreed with @alainfrisch was to follow 1 which is what the current state of PRs implement. 

1. The new modules `Float`,`Bool`, `Int` all have an `of_string : string -> t option` and no raising variant (`Float` introduced in the last major OCaml version is to be corrected along this line after a decision has been made, I offered to do it once these `Bool` and `Int` PRs are merged). This contradicts the interface of `{Int32,Int64,Nativeint}.of_string` and note: *only these* ones. A potential regularization of the interfaces for functor arguments regarding the `string -> t option` interface could be done later via `{Bool,Float,Int,Int64,Int32,Nativeint}.parse`. Note that this would mean there would be no regularization for raising variants.
1. *Only* `{Bool,Int}.of_string_opt : string -> t option` are introduced. This means that we immediately get regularity for the `string -> t option` interface and thus likely eschew the need for `X.parse` regularization and we forget about regularizing raising variants.
4. *Only* `{Bool,Int}.parse : string -> t option` is introduced and a regularization to be done later via `{Float,Int64,Int32,Nativeint}.parse` and we forget about regularizing raising variants.
5. *Only* `{Bool,Int}.option_of_string : string -> t option` is introduced and a regularization in the other modules to be done later via `{Float,Int,Int32,Int64,Nativeint}.option_of_string` and we forget about regularizing raising variants.
2. Both a `{Bool,Int}.of_string : string -> t` that raises `Failure` (please note that the current `bool_of_string` raises `Invalid_argument`, not `Failure`) and `{Bool,Int}.of_string_opt : string -> t option` are introduced. This entails immediate full regularity for both raising and non-raising variants w.r.t. the current `Float`, `Int32`, `Int64` and `Nativeint`.

A few random comments of mine. 

* The reason I'm favouring 1. is on the ground that the `Stdlib` is also the place where conventions can get established and I think the signal that should be given is for users to have that signature for their `of_string` functions as I don't find `of_string_opt` to be a great name, it really reads from a string option to me. I also have the *impression* that the `of_string` function is the one users will naturally fish for.
* In 5. I'm a little bit concerned about the potential for nasty bugs regarding the change in exception names between `Bool.of_string` and `bool_of_string` but repeating the design error in `Bool.of_string` would be totally silly in my opinion.
* Only 5. would regularize raising variants. I'm not too concerned about regularizing raising variants. Especially I don't find @drup's comment about still having raising variants for efficiency reasons compelling: I mostly see these `of_string` functions as toys for quick hacks. This because the syntax they parse for integers is foreign to most standards (e.g. the `_` to separate digits). If you need performance or whenever you implement standards you likely reimplement and specialize integer parsing yourself at which point you can use an exception for control flow yourself if you wish so.

alainfrisch commented:
After given it some thoughts, my preference would be for adding `*.parse` everywhere (new and existing modules), suggesting in the doc to use them rather than `*.of_string/*.of_string_opt` and eventually (i.e. not now) formally deprecating existing `*.of_string_opt` (without adding them for new modules).  As for `*.of_string` (raising variants), I'd rather not introduce them for new modules and rather focus efforts on making sure there is no performance degradation when calling `*.parse` and immediately pattern-matching on its result.  If this can be achieved, existing `*.of_string` should eventually be deprecated as well.

Drup commented:
I'm also okay with @alainfrisch's plan/number 3.

alainfrisch commented:
> I actually had the plan to work on this after these things are in. So if you really think it's not worth it please say.

One should be careful about not killing hopes of avoiding allocations when using option-returning variants. That said, the raising variant might be quite bad as well (even in the success case, because setting-up the trap is not free), and the solution might be some customized "unboxed" calling conventions, allowing a C primitive to return an option type without allocating the "Some" constructor.

mshinwell commented:
I'm unclear why it's proposed to change a name, `of_string`, which:
- is descriptive;
- has obvious symmetry with its counterpart (namely `to_string`);
- is widely used.

If the reason is for backwards compatibility, well, it doesn't seem sustainable to me in the longer term to never change the signatures of existing functions in the stdlib.  I would suggest what is required here is a policy, which is announced sufficiently far in advance, with some expectation on the part of users that they may need to update their code from time to time.  I expect within the next few years we will see automated refactoring tools that can help with this.

As for avoiding allocation, the next version of Flambda will be able to deal with the "option" cases, including across function returns.

alainfrisch commented:
@mshinwell  The problem with signature changes is that they cannot be communicated in advance in the form of a deprecation warning and that they cannot be anticipated by users before we actually break their code.  Which migration path do you suggest, in practice?

mshinwell commented:
I don't know exactly what the policy should be.  However I would imagine it would involve some specification as to when, presumably based on version numbers, changes of a certain form (for example deprecating a function or renaming a function in the stdlib) could be expected to occur.

I think the actual business of fixing code for things such as straightforward renames is probably actually rather less bad than its typical portrayal.  You can go an awfully long way just with `sed`.  Significant recent improvements to packaging tools and build systems are I think reducing the maintenance burden here too.

alainfrisch commented:
Such breaking changes are ok in some environment, typically corporate development with a synchronized upgrade to a newer version of OCaml, and no need to maintain compatibility with older versions.  But the story is less compelling for open-source projects.  In a sense, forcing maintainers of open-source packages to drop compatibility with older versions of OCaml could be seen as a way to move the community globally towards newer versions, for the benefits of everybody; but I'm not sure there is a consensus here.

alainfrisch commented:
Also, such breaking changes should not happen at every releases; and unless the next one officially becomes a breaking one, your suggestion means that we should either (1) delay the addition of the discussed features to the next version; (2) introduce features that we know we will break later.

dbuenzli commented:
> (1) delay the addition of the discussed features to the next version;

If there's disagreement on how to proceed I can simply remove `{Bool,Int}.of_string` from the PRs (and maybe retract it from `Float`). While `X.parse` is a solution that achieves our goal I can certainly relate to @mshinwell's comment [here](https://github.com/ocaml/ocaml/pull/2011#issuecomment-422781800).

dbuenzli commented:
Note however that again as far as I'm concerned I don't find it super dramatic that `{Int,Float,Bool}.of_string` use `option` but `{Int32,Int64,Nativeint}.of_string` raises for the next ten years. 

The stdlib has a history as long as OCaml and an inconsistency here and there is to be expected. I think we should rather strive to provide the good tools and the good names when we introduce new features rather than obsessing *too much* with consistency with a past we would like to get rid of.

dbuenzli commented:
The more I think about this, the more I'm convinced by my [last comment](https://github.com/ocaml/ocaml/pull/2011#issuecomment-422866023) and find the `parse` naming solution cumbersome -- `{of,to}_string` remains what most external libraries use and what new ones are likely to use. I think that introducing new terminology because *three* functions bear that name but don't have the right interface in the stdlib is a misguided step, especially since that new `parse` naming convention will likely only end up being used in the stdlib. 

As far as I'm concerned I will not be the person who introduces `parse` as I'm not convinced by its pertinence. Since the  `{Bool,Int}.of_string` functions are only a fraction of what the `Bool` and `Int` PR contain here are my proposals to unstuck the situation. Either:

1. I remove `{Bool,Int}.of_string` from the proposals, and if deemed reasonable and unless someone steps up to implement the `parse` route, also PR to remove the `Float.of_string` function.
2. There's enough agreement from the dev team members to proceed as currently implemented by the PRs and finish the work by making a PR to change the interface of the `Float.of_string` function (as suggested by @alainfrisch [here](https://github.com/ocaml/ocaml/pull/2010#issuecomment-420211054)).

dbuenzli commented:
Ping.

alainfrisch commented on stdlib/int.mli:
```diff
@@ -0,0 +1,140 @@
+    [0X]), in octal (if it begins with [0o] or [0O]), or in binary
+    (if it begins with [0b] or [0B]).
+
+    The [0u] prefix reads the input as an unsigned integer in the range
```
Some tweaks to the spec:

- The base prefixes come after the sign (- / +) (i.e. "beings with" is not accurate).
- 0U is also supported.
- The underscore cannot occur "in the prefix" (before the sign, between the sign and the base, in the middle of the base).
- 0u prefix allow negative numbers as well (upto something like -2*max_int-1).

Also, a design question: do you know why we cannot specify "unsigned" for non-decimal bases?

dbuenzli replied:
> Also, a design question: do you know why we cannot specify "unsigned" for non-decimal bases?

No Idea. I just copied that docstring from [`Stdlib.int_of_string_opt`](https://github.com/ocaml/ocaml/blob/1f25d352372ce6fe1cec5f862ec9c14222d79000/stdlib/stdlib.mli#L628-L647).

Will correct it as per your comments.

dbuenzli replied:
Done.

alainfrisch commented:
I'm now tempted to go for 2 (i.e. the PR in its current state, after tweaking the spec of `of_string`).  If someone objects to it, speak up now!

Drup commented:
@alainfrisch I do, for exactly the same reasons that I gave previously ....

Honestly, I'm very surprised that you are all ready to give up on having a consistent API so easily, especially for something as minor as a name that doesn't sound "exactly right". `parse` is a fine name, and there are other names we could chose. Having inconsistent API breaks the mental model and the expectations of users. The argument that it's done all the time in the stdlib should not encourage us to keep doing it, on the contrary.

Currently, `of_string` is indeed common in external libraries .... with a type `string -> _`. If we are to introduce a *new* standard, it is better to introduce a new name as well! That way, external libraries can be extended with a new function `parse : string -> _ option` and adopt this without breaking their APIs.

@alainfrisch If the only thing needed to converge to a decent API is for someone to volunteer to add `parse` everywhere, I can do that.

mshinwell commented:
Honestly, `of_string` is what everyone uses, so it's unclear to me why we should change that.  I agree with @dbuenzli that exact signature compatibility with older parts of the stdlib does not seem like a primary goal.  I think for a user it is worse being unable to have the function name to hand rather than having to make a quick adjustment when the type is not as expected.

Drup commented:
@mshinwell So what do you propose external libraries to do ? Slowly break lot's of API one by one as the improved signature is adopted throughout the ecosystem ? Adding a new name seem like such a minor thing by comparison.

alainfrisch commented:
> of_string is what everyone uses

I'm not familiar with alternative standard libraries, but do they typically use option-returning variants for their of_string functions?   (https://github.com/janestreet/base/blob/master/src/stringable.ml suggests that JS Base use exceptions)

Drup commented:
@alainfrisch Base and batteries use `of_string : string -> _`.

Edit: Apparently, containers uses `of_string : string -> _ option` for all modules, including the old ones.

alainfrisch commented:
Ok, I was operating under the assumption that everyone except the stdlib used option-returning variants instead of exceptions, following @mshinwell claim (which I misunderstood).  If this is not the case, I tend to agree with you, @Drup .

mshinwell commented:
(@alainfrisch To be clear, I meant "of_string" or an obvious variant of that name.  I was talking about the name rather than the exact signature.)

@Drup I'm not sure what you are referring to, because this pull request is proposing to add new functions with these names.  If you are talking about fixing the old functions in due course then, yes, some code fixing will have to be done (I made some comments on that further up).

Drup commented:
@mshinwell Let's be realistic here. Given the rate of breaking changes in the stdlib, those functions will never be fixed, since we can't propose a smooth migration path. Claiming that it will be done is pure fantasy unless you propose a concrete plan.

mshinwell commented:
I didn't claim that it will be done.

I will restate that it seems better to me to have functions with names that are familiar even if their type signatures are slightly unfamiliar (and occasionally inconsistent).  Think about discoverability with Merlin completion, for example.

alainfrisch commented:
Do Base/Core/Batteries/Containers have option-returning variants of of_string functions?

dbuenzli commented:
I have to admit that I'm getting tired of this discussion. 

I made the case that introducing `Bool.of_string` that raises `Failure` in 2018 is a very bad idea and I think the `stdlib` should show the good pattern which is to return options. Alternate stdlib are certainly a source of inspiration but I don't think they should dictate what should be done here.

So if @alainfrisch comes back on his decision I will simply remove the `of_string` function and let the people who have strong opinions about this implement what they see as being their perfect solution which I personally don't find compelling at all.

Drup commented:
@alainfrisch Base and Batteries do not. 

Containers for `Int32` has:
```ocaml
  val CCInt32.of_string_exn string -> t                                         
  val CCInt32.of_string string -> t option                                      
  val CCInt32.of_string_opt string -> t option 
```
and for `Int`:
```ocaml
val CCInt.of_string string -> t option
```

c-cube commented:
If you make `Int.of_string` return an option, please include a `Int.of_string_exn` or something similar to make migration of codebases easier. 

edit: also, at least for batteries and containers this is going to be a really painful change, probably involving a lot of cppo to keep a uniform API on old and new versions of OCaml.

dbuenzli commented:
> edit: also, at least for batteries and containers this is going to be a really painful change, probably involving a lot of cppo to keep a uniform API on old and new versions of OCaml.

@c-cube No signature is being changed. New functionality is being added.

c-cube commented:
Right, I guess I can just disable deprecation warnings. But does that mean that deprecated functions will never go away?

dbuenzli commented:
> Right, I guess I can just disable deprecation warnings. But does that mean that deprecated functions will never go away?

Nothing is being deprecated for now.

Chris00 commented:
I my opinion, coherence of the standard library is very important.  Not only is it easier for developers to use but incoherence make it appear as not having be well thought out which sheds a bad light on the language itself.

Since there does not seem to be a pain free way of updating functions and not breaking older packages, I think changes breaking backward compatibility should be grouped together as much as possible.  A way to proceed could be as follows.

- Launch a discussion about the principles that the standard library should embody, much in the same way as what was [done for Core](https://blog.janestreet.com/core-principles-uniformity-of-interface/).  For the particular point that makes much of the above discussion, I personally would be in favor of
    ```ocaml
    val of_string : string -> t option
    val of_string_exn : string -> t
    ```
  in all modules on the ground that exception raising functions for _non-exceptional behavior_ should be more visible as the programmer needs to exercise some caution when using them.  But then all `_opt` functions should be deprecated and the corresponding exception raising function renamed `_exn`.  For example `MapS.find: key -> 'a t -> 'a` becomes `MapS.find_exn: key -> 'a t -> 'a` and `MapS.find_opt: key -> 'a t -> 'a option` becomes `MapS.find: key -> 'a t -> 'a option`.
- Once these principle are agreed upon, the whole stdlib must be revised to implement them.  For the following, assume that we do this for OCaml 4.08.
- Have [stdlib-shims](https://github.com/ocaml/stdlib-shims) split into two branches: one will release the opam package `stdlib-shims 4.07` that contains the current library and the other one (the `master` branch) contains the revised version, release as `stdlib-shims 4.08`.
- People who want to use the new style with older compilers simply depend on `stdlib-shims 4.08`.
- For the already released packages, a automatic procedure should add to `depends` the line `"stdlib-shims" {= 4.07 & ocaml-version < 4.08}` and enrich (or create) `setenv` with
  ```
  [OCAMLPARAM = "-open stdlib_shims"]
  ```
  This point is a bit painful (since, after testing, some packages will require patches to be included in opam) and that's why I suggest to group these changes.

(If the constraint `ocaml-version < 4.08` is not supported, one can put the version number into the library and module names: say `stdlib_4_08`.)

pmetzger commented:
I note that in the past, labeled versions of standard library modules were created rather than breaking the compatibility of the existing standard library modules. Perhaps new standard library modules that embodied the new design principles could be created, and one could choose to use them or remain with the old ones in old code? (This is hardly an original thought of course.)

dbuenzli commented:
So now that @drup has shown us (thanks) that even the alternate standard libraries do not agree on what the right signature is for `of_string` functions could we at least move on with this PR according to one of the options [mentioned here](https://github.com/ocaml/ocaml/pull/2011#issuecomment-423892590). I'd really like to have the `Bool` and `Int` modules in 4.08 (and fix or remove `Float.of_string` if needed). 

I think the design space has been pretty thoroughly discussed and analysed at that point a summary of which can be found [here](https://github.com/ocaml/ocaml/pull/2011#issuecomment-422275177) (though this summary doesn't include the possibility of changing the signature of the only three `M.of_string` functions of the stdlib that raise as I thought that would be unacceptable).

alainfrisch commented:
I think dropping of_string from the PR should be pretty consensual.  It allows to move on and leave the tricky discussion/decision for later.

dbuenzli commented:
> I think dropping of_string from the PR should be pretty consensual. 

Ok I'll do this in the upcoming days. Should I then PR to remove `Float.of_string` aswell ? 

> It allows to move on and leave the tricky discussion/decision for later.

I can't help thinking that is simply shunning the issue. My impression that exactly the same tricky discussion/decision will occur later and we'll all lose our time a second time. But so be it.

alainfrisch commented:
My understanding is that stdlib-shims backports new features of the stdlib so that they can be used with older versions of OCaml (forward compatibility).  Hopefully, the community can come up with a nice solution for backward compatibility: allowing the upstream stdlib to introduce breaking changes without breaking existing packages (i.e. providing snapshots of older stdlib interfaces to be used with newer version of OCaml, making sure we keep enough type equalities to allow using multiple versions together).   Then the discussion will become easier (make option-returning variants the default across the entire stdlib).

Honestly, I sometimes wished the community put less energy on making sure people never need to upgrade to newer versions of OCaml and more energy on making it easier to move to newer versions...

dbuenzli commented:
> Then the discussion will become easier (make option-returning variants the default across the entire stdlib).

Maybe but to be honest I don't believe that much in these "hopefully the community".

Meanwhile we will have to explain to people why these modules don't have an `of_string` function while other in the stdlib have... Which of course is as much inconsistent as having `of_string` in all of the modules albeit with a few of these using options. Except the latter would still, as has been mentioned [by others](https://github.com/ocaml/ocaml/pull/2011#issuecomment-426991118), provide a better experience for the stdlib despite the slight operating inconsistency.

You didn't respond about removing `Float.of_string`.

alainfrisch commented:
Float.of_string:  perhaps not worth spending time discussing about its removal, I'd say.

(You seem unhappy with the approach I suggest, but 2 hours ago you were suggesting it was one of two alternatives you'd be ok to follow "we at least move on with this PR according to one of the options [mentioned here](https://github.com/ocaml/ocaml/pull/2011#issuecomment-423892590)", right?)

> Meanwhile we will have to explain to people why these modules don't have an of_string function while other in the stdlib have...

Easy, just refer them to https://github.com/ocaml/ocaml/pull/2011

dbuenzli commented:
> Float.of_string: perhaps not worth spending time discussing about its removal, I'd say.

Well not doing anything before 4.08 is basically casting a vote in favor of raising variants. ~~Since I no longer trust your leadership on that design issue~~ I will make a PR to start a discussion about it anyways.

> (You seem unhappy with the approach I suggest, but 2 hours ago you were suggesting it was one of two alternatives you'd be ok to follow "we at least move on with this PR according to one of the options [mentioned here](https://github.com/ocaml/ocaml/pull/2011#issuecomment-423892590)", right?)

There's nothing inconsistent here. I'm ok to follow it but yes I'm highly unhappy with the approach and just want to make people aware of the consequences as I don't think everyone realized that removing the function would be as inconsistent as having them. 

I find it a bit sad that the stubborn position of *one* individual basically brings us a more inconsistent and less useful stdlib. My feeling after all this is that there's no real interest to have a good stdlib within the design constraints that are given to us. Frankly individuals looking for the perfect one and are not able to think within the design constraints of the actual one should rather go work on their own rather than actively help in creating a broken stdlib.

alainfrisch commented:
I don't know which individual you are referring too, but : 
 - @Drup (https://github.com/ocaml/ocaml/pull/2011#issuecomment-426984809) insisted strongly on a uniform API 
 - @Chris00 (https://github.com/ocaml/ocaml/pull/2011#issuecomment-427603225) also claimed that "coherence of the standard library is very important"
 - I also find value in having a coherent API,  and I believe there exist at least two decent ways to reach that which deserve to be explored further (easy: introducing new names + deprecating old functions; harder: break compatibility and provide migration layers).

But your final point is spot on, so I'll unsubscribe from this PR.

gasche commented:
Come on, @dbuenzli, the handling of stdlib PRs is frustrating (for everyone), but your last message is rude to @alainfrisch, which is neither pleasant nor productive. It's a shitty job to be handling stdlib changes for reasons exemplified in this PR, and most people (myself included!) decided to *not* apply for it. @alainfrisch is currently the most active in processing stdlib proposals (with @nojb), and it's a bit sad that you would alienate him.

Hacking on the stdlib sucks, and we all know it, because there are many competing choices and often no clear "better choice" to agree on (in absence of a better choice, doing nothing is often the only reasonable-looking option). (Syntax discussions are often like that, except for *really good proposals* like `match ... with exception`). Sometimes one wins an argument, sometimes everyone loses, and I think we should try not to be too harsh about it.

dbuenzli committed:
Remove Int.of_string.

dbuenzli commented:
> Come on, @dbuenzli, the handling of stdlib PRs is frustrating (for everyone), but your last message is rude to @alainfrisch, which is neither pleasant nor productive.

I apologize to @alainfrisch for it. `Int.of_string` has been removed.

mshinwell commented:
I can't quite believe that it is now proposed to add these modules without the `of_string` functions.  I agree entirely with the second paragraph of [this comment](https://github.com/ocaml/ocaml/pull/2011#issuecomment-427991926).

Bear in mind that unless you're writing a library from scratch, having already written such a library before with all the pitfalls and potential inconsistencies already in your head, then there is always going to be imperfection.  Yes, we should have policies about breaking backwards compatibility; yes, we should have good refactoring tools; and yes, older function signatures should be updated so that everything is consistent.  However blocking the addition of new, useful functions that have standard, discoverable names and _just slightly different_ type signatures because we haven't yet solved those problems seems the wrong decision to me.  Many users probably won't even notice these minor inconsistencies on a large scale---it will be a minor irritation now and then.

The only real question to my mind that needed to be answered here is what the new standard interface for the functions in question should be (it would seem wrong to add new functions with old signatures, especially if such signatures are problematic in some way as the ones in question potentially are).  That seems like it could have been worked out rather than kicking it into the long grass.

alainfrisch commented:
(Ok, unsubscribe is not enough, since explicit @-mentions re-subscribe automatically.)

> However blocking the addition of new, useful functions that have standard

It's really not like this is blocking the addition of new useful stuff:  all suggest features are already part of the Sdtlib and `Stdlib.int_of_string_opt` follows the current naming convention in the Stdlib.  The main practical value of the Int module, in my opinion, is to allow e.g. `Set.Make(Int)` (btw, why not add Int.hash?), and to provide a namespace for future, more specialized, additions (for instance a popcount function).  Another value of the PR is to create uniformity, but since we cannot fully easily achieve it now, at least let's not make it more difficult to get it later.

The decision of switching to option-returning variants pervasively in the Stdlib, for all conversion and lookup functions, is a topic on its own, rather independent of the addition of Int and much less consensual.

Drup commented:
> I find it a bit sad that the stubborn position of _one_ individual basically brings us a more inconsistent and less useful stdlib. My feeling after all this is that there's no real interest to have a good stdlib within the design constraints that are given to us. Frankly individuals looking for the perfect one and are not able to think within the design constraints of the actual one should rather go work on their own rather than actively help in creating a broken stdlib.

I suppose you mean yourself, by camping on your position and refusing to add the basic `of_string : string -> _` function to stdlib, in favor of the fancier and more modern `string -> _ option` variant, or even the slightly less elegant `parse` name.

Let's not play silly games @dbuenzli . Don't paint me as the bad guy here, we all have our pet-peeve but at least I recognized the value in yours and I was ready to compromise. I still think `parse` is a perfectly good solution.

mshinwell commented:
@Drup It has been pointed out that there are good reasons for preferring a standard name, also, though.

Did you mean "`of_string`" in your example?
Assuming that is what you meant, then I do sense some aversion to the option-returning versions.  I would point out the following here: these kinds of conversion functions are a source of errors.  A library should be guarding against such errors as much as it reasonably can.  As an example, I've been working recently with a lot of conversions between different integer sizes (some related to cross compilation issues), and it's really painful to have to deal with conversion functions that cannot return results as `option` or equivalent.  This isn't just a matter of being "modern", it's also a matter of having the right functions to write robust code.  If you're going to be ideologically set on a particular position, as some of our contributors may or may not be, it doesn't seem so bad to come down on the side of an `option`-returning conversion being the default.

Drup commented:
@mshinwell As my previous messages on the matter should make clear, I'm not against the option returning at all. But either you do it properly (by changing everything and breaking compat) or you do it differently (by adding an extra function with that api, like `parse`, and setting a new standard which I'll happily adopt in my libraries). I'm against the solution which consist of doing something inconsistent that doesn't fix the problem properly.

dbuenzli commented:
> I suppose you mean yourself, by camping on your position and refusing to add the basic `to_string : _ -> string` function to stdlib, in favor of the fancier and more modern `_ -> string option` variant, or even the slightly less elegant `print` name.

Please I was not alone. Moreover you said you were willing to compromise but you never responded with any good argument to the concern of introducing a `Bool.of_string` function raising `Failure _` while `bool_of_string` has been raising `Invalid_argument` for ages.

Anyways that discussion is over for me. I only commented the function since I did spend sometime on improving the doc string on `int` parsing. If people want to revert that decision at some point it will be easy to do.

bluddy commented:
I do think people tend to associate a name - especially a simple one - with a given type signature template. Remembering that `Int.of_string` has an option return type while other `of_string` functions raise exceptions is going to be really hard once we forget that `Int` was even added later, which will only take a couple of versions. It'll be something that only the old-timers remember, and even then we won't be sure how it happened and need to consult the resident historian.

So since we all basically agree that we like option-returning functions and dislike exception-based ones, why don't we incorporate `of_string` here with exceptions to be consistent, and then work on a global PR that renames common exception-raising functions and deprecates the old names? We can do a PR per function or per a few functions. Once we have the new names, any new modules can immediately use the new names and types.

Chris00 commented:
> A library should be guarding against such errors as much as it reasonably can. […] write robust code.

I agree with that too, so these should be the default.  Occasionally, functions raising exceptions are convenient and a suffix `_exn` reminds the reader of their behavior.  Moreover, I would argue that we need to use standard names.  So the only way forward it to break backward compatibility.  As already mentioned, this is going to be a bit painful, so we have to propose a carefully thought migration path (I tentatively proposed one above but nobody commented on it) and preferably do it only once (which is why I think some design principles have to be explicitly be laid out).  Many alternatives to the standard library have been proposed over the years and no one has “won” — and despite all the energy put into these, I'm not convinced that it will happen as it did in the package and the build departments.  Also, from the perspective of a newcomer, the message that the _standard library_ coming with the compiler is better replaced with another one may be worrisome.  So I believe that _stdlib_ should use all the experience gathered from alternatives and showcase the highest quality, consistency, and best practices.

dbuenzli commented:
> a carefully thought migration path (I tentatively proposed one above but nobody commented on it)

The problem with your proposal is that it doesn't entice people to upgrade and it's also quite heavy both in terms of maintenance and build obfuscation (`OCAMLPARAM`). 

I think the only sane approach for the `stdlib` is to try to *gradually* improve it, with occasional hard and well thought out break points. Even if this involves, meanwhile, transient, non-harmful (because they establish better standards and do not break code), inconsistencies *of certain kind*. 

Maybe what this discussion failed at to point out is that that there are 1) different kind of inconsistencies and 2) these have different properties, from harmless to harmful and 3) that the same holds for *consistencies*. The inconsistency that I tried to introduce maybe felt embarrassing to some but it was not harmful from a programming perspective (amusingly the end-result is as embarrassing now).

Software is never perfect, I'm routinely dissatisfied with designs I started from scratch and thought to be the ultimate one when I created them. Errors are and will still be made even in new interfaces that are being introduced now. Asking for perfection, especially in a project like the OCaml stdlib is just aiming to eventually do nothing and perpetuate our dissatisfaction with the stdlib. 

I think that all or nothing attitudes like [that](https://github.com/ocaml/ocaml/pull/2011#issuecomment-428132420) (I'm pointing the attitude not the person) will make it difficult to make significant progress on the stdlib in the future. Gradual improvement with a vision and end-user onboarding is the only solution in my opinion.

Chris00 commented:
> The problem with your proposal is that it doesn't entice people to upgrade and it's also quite heavy both in terms of maintenance and build obfuscation (OCAMLPARAM).

As I see it, there is _no_ maintenance — this solution is for _existing_ packages in the repo to compile with the newer compilers.  I did not mean to do this for new packages submitted to the repo (which can adjust their build systems).

dbuenzli commented:
> As I see it, there is _no_ maintenance —

I meant maintenance around providing the compatibility layer. Also it may not always be possible to provide one (e.g. when new language features/primitives are introduced). 

>  this solution is for _existing_ packages in the repo to compile with the newer compilers.

That's precisely what I think should be avoided. When you move the compilers and break things you should entice the eco-system to update and move with you rather than allow it to function with older assumptions.

Chris00 commented:
> move with you rather than allow it to function with older assumptions.

The only way I see then is that _stdlib_ be released as separate packages (with a version number in the name since there will be a need to install several of them concurrently).  One just increases the version number if there is backward compatibility and release a new package if not?  _ocaml_ packages could just be meta-packages depending on _ocaml-compiler_ and _stdlib_ (in order to keep the fact that they are blessed because they come with the compiler).

dbuenzli commented:
I think the single remaining objection to this PR has been solved by https://github.com/ocaml/ocaml/pull/2011/commits/6a94394cb858b3b53a0a3fdf00d77b5d1cb67753 and this can be merged.

nojb commented:
That being the only point of disagreement, I went ahead and merged.

Thanks for your efforts and perseverance!

dbuenzli commented:
Thanks @nojb !

Drup commented:
@dbuenzli Despite my oppositions, I appreciate you volunteered to do those PRs. Thanks!

dbuenzli commented:
Thanks @drup and sorry again to have lost my temper, especially to @alainfrisch.

alainfrisch commented:
Sure, no problem @dbuenzli.
